Wrong inference of dataclasses with custom __init__ and varargs.

Small case to reproduce:

```python
from typing import Tuple
from dataclasses import dataclass

@dataclass
class Foo:
    spams: Tuple[int, ...]

    def __init__(self, *spams: int) -> None:
        self.spams = spams

def build_foo(*spams: int) -> Foo:
    return Foo(*spams)
```

Checking that program with mypy 0.670 (Python 3.6), results in:

```
bug.py:13: error: Argument 1 to "Foo" has incompatible type "*Tuple[int, ...]"; expected "Tuple[int, ...]"
```

Without the `@dataclass` decorator the program passes.  It seems that the dataclass is hiding the actual ` __init__`.

--- a/mypy/plugins/dataclasses.py
+++ b/mypy/plugins/dataclasses.py
@@ -92,7 +92,11 @@ class DataclassTransformer:
             'frozen': _get_decorator_bool_argument(self._ctx, 'frozen', False),
         }
 
-        if decorator_arguments['init']:
+        # If there are no attributes, it may be that the new semantic analyzer has not
+        # processed them yet. In order to work around this, we can simply skip generating
+        # __init__ if there are no attributes, because if the user truly did not define any,
+        # then the object default __init__ with an empty signature will be present anyway.
+        if decorator_arguments['init'] and '__init__' not in info.names and attributes:
             add_method(
                 ctx,
                 '__init__',

--- a/test-data/unit/check-dataclasses.test
+++ b/test-data/unit/check-dataclasses.test
@@ -16,6 +16,21 @@ Person('Jonh', 21, None)  # E: Too many arguments for "Person"
 
 [builtins fixtures/list.pyi]
 
+[case testDataclassesCustomInit]
+# flags: --python-version 3.6
+from dataclasses import dataclass
+
+@dataclass
+class A:
+    a: int
+
+    def __init__(self, a: str) -> None:
+        pass
+
+A('1')
+
+[builtins fixtures/list.pyi]
+
 [case testDataclassesBasicInheritance]
 # flags: --python-version 3.6
 from dataclasses import dataclass

